src/which.js

```diff
@@ -45,74 +39,67 @@ function _which(options, cmd) {
 
   var pathEnv = process.env.path || process.env.Path || process.env.PATH;
   var pathArray = splitPath(pathEnv);
-  var all = options.all;
-  var where = all ? [] : null;
+
+  // An object to store the matches for the `which` query
+  var queryMatches = {
+    add: function (file) { this.value.push(file); },
+    // Return the first result only unless options.all is specified
+    getValue: function () {
+      return options.all ? this.value : this.value[0];
+    },
+    hasFoundSomething: function () { return Boolean(this.value.length); },
+    // This is the value to return if we haven't found any matches
+    noMatches: options.all ? [] : null,
+    value: [],
+  };
```

Comment:
This is better than it was before, but this is still a little complicated. On the one hand, this generalizes the two options more clearly, but on the other hand, dealing with one-off objects is a bit messy.